Git - Ocasionally Commit & Push fails with "Actual command not found"

Type: <b>Bug</b>

Ocasionally I face the following error

1. Have pending changes
2. Click  Commit & Push
3. See error notification
4. Next attempt works without issues

![image](https://github.com/microsoft/vscode/assets/145683618/446f22b1-b471-497d-9512-5e7e688fdd2b)

Developer Tools

```
10:54:07.622 notificationsAlerts.ts:40 Error: Actual command not found, wanted to execute git.commit /30
    at r.h (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:153:189812)
c @ notificationsAlerts.ts:40
```
GIT Output

```
2024-05-07 10:54:04.887 [info] > git config --get commit.template [94ms]
2024-05-07 10:54:04.956 [info] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track)%00%(upstream:remotename)%00%(upstream:remoteref) --ignore-case refs/heads/feature/<REDACTED> refs/remotes/feature/<REDACTED> [130ms]
2024-05-07 10:54:05.090 [info] > git status -z -uall [118ms]
2024-05-07 10:54:05.236 [info] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track)%00%(upstream:remotename)%00%(upstream:remoteref) --ignore-case refs/heads/feature/<REDACTED> [114ms]
2024-05-07 10:54:05.380 [info] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track)%00%(upstream:remotename)%00%(upstream:remoteref) --ignore-case refs/heads/feature/<REDACTED> [119ms]
2024-05-07 10:54:05.531 [info] > git config --local branch.feature/<REDACTED>.vscode-merge-base [94ms]
2024-05-07 10:54:05.531 [warning] git config failed: Failed to execute git
2024-05-07 10:54:05.647 [info] > git reflog refs/heads/feature/<REDACTED> --grep-reflog=branch: Created from *. [95ms]
```

VS Code version: Code 1.89.0 (b58957e67ee1e712cebf466b995adf4c5307b2bd, 2024-05-01T02:09:22.859Z)
OS version: Windows_NT x64 10.0.22631
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|12th Gen Intel(R) Core(TM) i7-1270P (16 x 2496)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|31.67GB (11.14GB free)|
|Process Argv|--folder-uri file:///c%3A/sources/Repos/FLT --crash-reporter-id 0ed2dc76-86cb-4e82-89ae-a98b71ec0a3f|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (15)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-tailwindcss|bra|0.10.5
copilot|Git|1.187.0
copilot-chat|Git|0.15.0
vscode-github-actions|git|0.26.2
rainbow-csv|mec|3.11.0
vscode-docker|ms-|1.29.1
vscode-dotnet-runtime|ms-|2.0.3
vscode-edge-devtools|ms-|2.1.5
data-workspace-vscode|ms-|0.5.0
mssql|ms-|1.22.1
sql-bindings-vscode|ms-|0.4.0
sql-database-projects-vscode|ms-|1.4.1
live-server|ms-|0.4.13
powershell|ms-|2024.2.1
vscode-thunder-client|ran|2.22.1


</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vscorecescf:30445987
vscod805:30301674
binariesv615:30325510
vsaa593:30376534
py29gd2263:31024239
vscaat:30438848
c4g48928:30535728
azure-dev_surveyonecf:30548226
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pyreplss1:30897532
pythonmypyd1:30879173
pythoncet0:30885854
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
ccp2r3:30993541
dsvsc020:30976470
pythonait:31006305
showvideot:31016892
chatpanelc:31018788
dsvsc021:30996838
724cj586:31013169
pythoncenvpt:31022790
fchga941:31038042
dwnewjupytercf:31035177

```

</details>

<!-- generated by issue reporter -->

--- a/src/vs/workbench/api/common/extHostSCM.ts
+++ b/src/vs/workbench/api/common/extHostSCM.ts
@@ -693,7 +693,9 @@ class ExtHostSourceControl implements vscode.SourceControl {
 		this.#proxy.$updateSourceControl(this.handle, { acceptInputCommand: internal });
 	}
 
-	private readonly _actionButtonDisposables = new MutableDisposable<DisposableStore>();
+	// We know what we're doing here:
+	// eslint-disable-next-line local/code-no-potentially-unsafe-disposables
+	private _actionButtonDisposables = new DisposableStore();
 	private _actionButton: vscode.SourceControlActionButton | undefined;
 	get actionButton(): vscode.SourceControlActionButton | undefined {
 		checkProposedApiEnabled(this._extension, 'scmActionButton');
@@ -710,26 +712,32 @@ class ExtHostSourceControl implements vscode.SourceControl {
 			return;
 		}
 
+		// In order to prevent disposing the action button command that are still rendered in the UI
+		// until the next UI update, we ensure to dispose them after the update has been completed.
+		const oldActionButtonDisposables = this._actionButtonDisposables;
+		this._actionButtonDisposables = new DisposableStore();
+
 		this._actionButton = actionButton;
-		this._actionButtonDisposables.value = new DisposableStore();
 
 		const actionButtonDto = actionButton !== undefined ?
 			{
 				command: {
-					...this._commands.converter.toInternal(actionButton.command, this._actionButtonDisposables.value),
+					...this._commands.converter.toInternal(actionButton.command, this._actionButtonDisposables),
 					shortTitle: actionButton.command.shortTitle
 				},
 				secondaryCommands: actionButton.secondaryCommands?.map(commandGroup => {
-					return commandGroup.map(command => this._commands.converter.toInternal(command, this._actionButtonDisposables.value!));
+					return commandGroup.map(command => this._commands.converter.toInternal(command, this._actionButtonDisposables));
 				}),
 				enabled: actionButton.enabled
-			} satisfies SCMActionButtonDto : undefined;
+			} satisfies SCMActionButtonDto : null;
 
-		this.#proxy.$updateSourceControl(this.handle, { actionButton: actionButtonDto ?? null });
+		this.#proxy.$updateSourceControl(this.handle, { actionButton: actionButtonDto })
+			.finally(() => oldActionButtonDisposables.dispose());
 	}
 
-
-	private readonly _statusBarDisposables = new MutableDisposable<DisposableStore>();
+	// We know what we're doing here:
+	// eslint-disable-next-line local/code-no-potentially-unsafe-disposables
+	private _statusBarDisposables = new DisposableStore();
 	private _statusBarCommands: vscode.Command[] | undefined = undefined;
 
 	get statusBarCommands(): vscode.Command[] | undefined {
@@ -741,12 +749,17 @@ class ExtHostSourceControl implements vscode.SourceControl {
 			return;
 		}
 
-		this._statusBarDisposables.value = new DisposableStore();
+		// In order to prevent disposing status bar commands that are still rendered in the UI
+		// until the next UI update, we ensure to dispose them after the update has been completed.
+		const oldStatusBarDisposables = this._statusBarDisposables;
+		this._statusBarDisposables = new DisposableStore();
 
 		this._statusBarCommands = statusBarCommands;
 
-		const internal = (statusBarCommands || []).map(c => this._commands.converter.toInternal(c, this._statusBarDisposables.value!)) as ICommandDto[];
-		this.#proxy.$updateSourceControl(this.handle, { statusBarCommands: internal });
+		const internal = (statusBarCommands || []).map(c => this._commands.converter.toInternal(c, this._statusBarDisposables)) as ICommandDto[];
+
+		this.#proxy.$updateSourceControl(this.handle, { statusBarCommands: internal })
+			.finally(() => oldStatusBarDisposables.dispose());
 	}
 
 	private _selected: boolean = false;

